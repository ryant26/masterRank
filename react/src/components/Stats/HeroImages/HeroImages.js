import React from 'react';
import Proptypes from 'prop-types';

import DisableableHeroImage from '../../Images/DisableableHeroImage/DisableableHeroImage';
import HeroImage from '../../Images/HeroImage/HeroImage';

const HeroImages = ({heroNames, disabled}) => {

    let heroImages = heroNames.map((name, i) => {
        return ( disabled[i]
            ? ( <DisableableHeroImage
                    key={i}
                    heroName={name}
                    disabled={true}
                />)
            : ( <HeroImage
                 key={i}
                 heroName={name}
                />)
        );
    });

    return (
        <div className="HeroImages flex">
            { heroImages }
        </div>
    );
};

HeroImages.defaultProps = {
    disabled: []
};

HeroImages.propTypes = {
    heroNames: Proptypes.array.isRequired,
    disabled: Proptypes.arrayOf(Proptypes.bool)
};

export default HeroImages;