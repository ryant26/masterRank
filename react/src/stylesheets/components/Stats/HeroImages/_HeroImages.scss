.HeroImages {

  .HeroImage {
    @include square;
    margin-right: 10px;
  }

  .disabled {
    @include square;
    margin-right: 10px;
  }
}