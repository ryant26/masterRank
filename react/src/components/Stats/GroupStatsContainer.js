import React from 'react';
import PropTypes from 'prop-types';

import HeroStatsList from './HeroStatsList/HeroStatsList';
import HeroImages from './HeroImages/HeroImages';
import Model from '../../model/model';

const GroupStatsContainer = ({group, isLeading, toggleModal}) => {
    const groupHeroes = [group.leader, ...group.members];

    const wins = groupHeroes.reduce((wins, hero) => {
        if (hero.stats && hero.stats.wins) {
            return wins + hero.stats.wins;
        }
        return wins;
    }, 0);

    const totalGames = groupHeroes.reduce((games, hero) => {
        if (hero.stats && hero.stats.gamesPlayed) {
            return games + hero.stats.gamesPlayed;
        }

        return games;
    }, 0);

    const groupWinRate = (totalGames ? wins / totalGames : 0) * 100;
    const groupSr = Math.floor(groupHeroes.reduce((sr, hero) => sr + hero.skillRating, 0) / groupHeroes.length);

    const leaveGroup = () => {
        toggleModal();
        let preferredHeroName = group.leader.heroName;
        Model.leaveGroup();
        Model.createNewGroup(preferredHeroName);
    };

    return (
        <div className="GroupStatsContainer">
            <div className="header">
                <div className="flex justify-between align-center">
                    <div>
                        <div className="title flex align-center">
                            <h3>{isLeading ? 'Your ' : `${group.leader.platformDisplayName}'s`}</h3>
                            <h3 className="regular">Group</h3>
                        </div>
                        <span className="sub-title"><b>{groupSr}</b> Group SR</span>
                        {groupWinRate ? <span className="sub-title"><b>{groupWinRate.toFixed(1)}% </b> Group Winrate</span> : ''}
                    </div>
                    <HeroImages heroNames={groupHeroes.map((hero) => hero.heroName)}/>
                </div>
            </div>
            <div className="body">
                <HeroStatsList
                    heroes={groupHeroes}
                    emptyMessage="We have no stats on this group."
                    showPlatformDisplayName={true}
                    groupLeader={group.leader.platformDisplayName}
                />
            </div>
            {!isLeading ?
                <div className="footer flex justify-between align-center">
                    <div className="sub-title">
                        <b>Players joined: {groupHeroes.length}</b>
                    </div>
                    <div className="button-six flex align-center" onClick={leaveGroup}>
                        <div className="button-content">
                            LEAVE GROUP
                        </div>
                    </div>
                </div> :
                undefined
            }
        </div>
    );
};

GroupStatsContainer.propTypes = {
    group: PropTypes.object.isRequired,
    isLeading: PropTypes.bool.isRequired,
    toggleModal: PropTypes.func
};


export default GroupStatsContainer;