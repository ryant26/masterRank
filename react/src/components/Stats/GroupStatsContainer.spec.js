import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import GroupStatsContainer from './GroupStatsContainer';
import HeroImages from './HeroImages/HeroImages';
import Model from '../../model/model';

import groups from '../../resources/groupInvites';

const shallowGroupStatsContainer = (group, leader, toggleModal) => {
    return shallow(
        <GroupStatsContainer group={group} isLeading={leader} toggleModal={toggleModal}/>
    );
};

describe('GroupStatsContainer Component', () => {
    let wrapper;
    let group;

    beforeEach(() => {
        group = groups[0];
        wrapper = shallowGroupStatsContainer(group, false);
    });

    it('should render without exploding', () => {
        expect(wrapper.find(GroupStatsContainer)).toBeTruthy();
    });

    it('should match the snapshot', () => {
        let component = renderer.create(
            <GroupStatsContainer group={group} isLeading={false}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it('should have the title your group if you are leading', () => {
        wrapper = shallowGroupStatsContainer(group, true);
        expect(wrapper.find('.title > h3').first().text().trim()).toEqual('Your');
    });

    it('should not show leave button if you are leading', () => {
        wrapper = shallowGroupStatsContainer(group, true);
        expect(wrapper.find('.button-six').length).toBeFalsy();
    });

    it('should show leave button if you are not leading', () => {
        expect(wrapper.find('.button-six').length).toBeTruthy();
    });

    it("should show the leader's hero images first", () => {
        expect(wrapper.find(HeroImages).props().heroNames[0]).toEqual(group.leader.heroName);
    });

    it('should calculate group SR correctly', () => {
       expect(wrapper.find('span.sub-title > b').first().text()).toEqual('2700');
    });

    it('should render group SR as a whole number', () => {
       group.leader.skillRating = 3201.99999;
       wrapper = shallowGroupStatsContainer(group, false);
       expect(wrapper.find('span.sub-title > b').first().text()).toEqual('2700');
    });

    describe('when button-six is clicked', () => {
        let toggleModal;

        beforeEach(() => {
            Model.leaveGroup = jest.fn();
            Model.createNewGroup = jest.fn();
            toggleModal = jest.fn();
            wrapper = shallowGroupStatsContainer(group, false, toggleModal);

            expect(Model.leaveGroup).not.toHaveBeenCalled();
            expect(Model.createNewGroup).not.toHaveBeenCalled();
            expect(toggleModal).not.toHaveBeenCalled();
            wrapper.find('.button-six').simulate('click');
        });

        it('should call Model.leaveGroup', () => {
            expect(Model.leaveGroup).toHaveBeenCalledWith(group.groupId);
        });

        it('should call Model.createNewGroup', () => {
            expect(Model.createNewGroup).toHaveBeenCalledWith(group.leader.heroName);
        });

        it('should call toggleModal', () => {
            expect(toggleModal).toHaveBeenCalled();
        });
    });
});