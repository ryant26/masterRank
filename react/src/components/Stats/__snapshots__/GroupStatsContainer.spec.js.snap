// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupStatsContainer Component should match the snapshot 1`] = `
<div
  className="GroupStatsContainer"
>
  <div
    className="header"
  >
    <div
      className="flex justify-between align-center"
    >
      <div>
        <div
          className="title flex align-center"
        >
          <h3>
            luckbomb#1234's
          </h3>
          <h3
            className="regular"
          >
            Group
          </h3>
        </div>
        <span
          className="sub-title"
        >
          <b>
            2700
          </b>
           Group SR
        </span>
        
      </div>
      <div
        className="HeroImages flex"
      >
        <img
          alt="genji icon"
          className="HeroImage"
          onClick={undefined}
          src={
            Object {
              "process": [Function],
            }
          }
        />
        <img
          alt="widowmaker icon"
          className="HeroImage"
          onClick={undefined}
          src={
            Object {
              "process": [Function],
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="body"
  >
    <div
      className="HeroStatsList"
    >
      <div
        className="HeroStatsListItem"
      >
        <div
          className="flex align-center"
        >
          <div>
            <img
              alt="leader-icon"
              className="crown"
              src={
                Object {
                  "process": [Function],
                }
              }
            />
            <img
              alt="genji icon"
              className="HeroImage"
              onClick={undefined}
              src={
                Object {
                  "process": [Function],
                }
              }
            />
          </div>
          <div>
            <h3>
              luckbomb#1234 - 
              Genji
            </h3>
            <div
              className="sub-title"
            >
               hours played
            </div>
          </div>
          <div
            className="flex justify-between record"
          >
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0
              </div>
              <div
                className="sub-title"
              >
                WINS
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0
              </div>
              <div
                className="sub-title"
              >
                LOSSES
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0.00
              </div>
              <div
                className="sub-title"
              >
                K/D
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex wrap justify-between"
        >
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              DAMAGE
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              HEALING
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              BLOCKED
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. KILLS
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. TIME
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              ACCURACY
            </div>
          </div>
        </div>
      </div>
      <div
        className="HeroStatsListItem"
      >
        <div
          className="flex align-center"
        >
          <div>
            <img
              alt="widowmaker icon"
              className="HeroImage"
              onClick={undefined}
              src={
                Object {
                  "process": [Function],
                }
              }
            />
          </div>
          <div>
            <h3>
              wismo0#1234 - 
              Widowmaker
            </h3>
            <div
              className="sub-title"
            >
               hours played
            </div>
          </div>
          <div
            className="flex justify-between record"
          >
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0
              </div>
              <div
                className="sub-title"
              >
                WINS
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0
              </div>
              <div
                className="sub-title"
              >
                LOSSES
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                0.00
              </div>
              <div
                className="sub-title"
              >
                K/D
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex wrap justify-between"
        >
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              DAMAGE
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              HEALING
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              BLOCKED
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. KILLS
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. TIME
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              ACCURACY
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer flex justify-between align-center"
  >
    <div
      className="sub-title"
    >
      <b>
        Players joined: 
        2
        /
        4
      </b>
    </div>
    <div
      className="button-six flex align-center"
      onClick={[Function]}
    >
      <div
        className="button-content"
      >
        LEAVE GROUP
      </div>
    </div>
  </div>
</div>
`;
